File: Semester 4/Programming and Projection Enironments/Proiect_MPP/src/main/java/root/proiect_mpp/repositories/people/employees/EmployeeRepository.java
package root.proiect_mpp.repositories.people.employees;

import root.proiect_mpp.domain.people.Employee;
import root.proiect_mpp.repositories.people.PersonRepository;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Collection;
import java.util.Properties;

public class EmployeeRepository extends PersonRepository<Employee> implements IEmployeeRepository {
    public EmployeeRepository(Properties properties) {
        super(properties);
        logger.info("Initializing Employee Repository ");
    }

    @Override
    protected String getTableName() {
        return "employees";
    }

    @Override
    protected Employee extractEntity(ResultSet resultSet) throws SQLException {
        int id = resultSet.getInt("id");
        String firstName = resultSet.getString("first_name");
        String lastName = resultSet.getString("last_name");
        String position = resultSet.getString("position");
        String email = resultSet.getString("email");
        String password = resultSet.getString("password");
        return new Employee(id, firstName, lastName, position, email, password);
    }

    @Override
    protected PreparedStatement createInsertStatement(Connection connection, Employee entity) throws SQLException {
        PreparedStatement statement = connection.prepareStatement(super.sqlCommand);
        // statement.setInt(1, entity.getId());
        statement.setString(1, entity.getFirstName());
        statement.setString(2, entity.getLastName());
        statement.setString(3, entity.getPosition());
        statement.setString(4, entity.getEmail());
        statement.setString(5, entity.getPassword());
        return statement;
    }

    @Override
    protected PreparedStatement createUpdateStatement(Connection connection, Employee entity, Integer integer) throws SQLException {
        PreparedStatement statement = connection.prepareStatement(super.sqlCommand);
        statement.setInt(6, entity.getId());
        statement.setString(1, entity.getFirstName());
        statement.setString(2, entity.getLastName());
        statement.setString(3, entity.getPosition());
        statement.setString(4, entity.getEmail());
        statement.setString(5, entity.getPassword());
        return statement;
    }

    @Override
    protected PreparedStatement createDeleteStatement(Connection connection, Employee entity) throws SQLException {
        PreparedStatement statement = connection.prepareStatement(super.sqlCommand);
        statement.setInt(1, entity.getId());
        return statement;
    }

    @Override
    public void add(Employee elem) {
        super.sqlCommand = "INSERT INTO employees (first_name, last_name, position, email, password) VALUES (?, ?, ?, ?, ?)";
        super.add(elem);
    }

    @Override
    public void delete(Employee elem) {
        super.sqlCommand = "DELETE FROM employees WHERE id=?";
        super.delete(elem);
    }

    @Override
    public void update(Employee elem, Integer integer) {
        super.sqlCommand = "UPDATE employees SET first_name=?, last_name=?, position=?, email=?, password=? WHERE id=?";
        super.update(elem, integer);
    }

    @Override
    public Employee findById(Integer integer) {
        super.sqlCommand = "SELECT * FROM employees WHERE id=" + integer;
        return super.getOne();
    }

    @Override
    public Employee findByEmail(String email) {
        super.sqlCommand = "SELECT * FROM employees WHERE email='" + email + '\'';
        return super.getOne();
    }

    @Override
    public Collection<Employee> getAll() {
        super.sqlCommand = "SELECT * FROM employees";
        return super.getAll();
    }

    @Override
    public Collection<Employee> getByPosition(String position) {
        super.sqlCommand = "SELECT * FROM employees WHERE position='" + position + '\'';
        return super.getAll();
    }
}

File: Semester 4/Programming and Projection Enironments/mpp-proiect-csharp/Proiect_MPP/Proiect_MPP/repository/people/employees/EmployeeRepository.cs
using log4net;
using Proiect_MPP.databaseUtils;
using Proiect_MPP.domain.people;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Proiect_MPP.repository.people.employees
{
    public class EmployeeRepository : PersonRepository<Employee>, IEmployeeRepository
    {

        public EmployeeRepository(IDictionary<string, string> properties) : base(properties)
        {
            logger.Info("Initializing Employee Repository.");
        }

        protected override string getTableName()
        {
            return "employees";
        }

        protected override Employee extractEntity(IDataReader dataReader)
        {
            int id = dataReader.GetInt32(0);
            string firstName = dataReader.GetString(1);
            string lastName = dataReader.GetString(2);
            string position = dataReader.GetString(3);
            string email = dataReader.GetString(4);
            string password = dataReader.GetString(5);
            return new Employee(id, firstName, lastName, position, email, password);
        }

        public override int add(Employee item)
        {
            base.connection = DbUtils.getConnection(base.properties);
            using (base.sqlCommand = connection.CreateCommand())
            {
                base.sqlCommand.CommandText = "INSERT INTO employees (first_name, last_name, position, email, password)" +
                    " VALUES (@firstName, @lastName, @position, @email, @password)";

                var paramFirstName = base.sqlCommand.CreateParameter();
                paramFirstName.ParameterName = "@firstName";
                paramFirstName.Value = item.FirstName;
                base.sqlCommand.Parameters.Add(paramFirstName);

                var paramLastName = base.sqlCommand.CreateParameter();
                paramLastName.ParameterName = "@lastName";
                paramLastName.Value = item.LastName;
                base.sqlCommand.Parameters.Add(paramLastName);

                var paramPosition = base.sqlCommand.CreateParameter();
                paramPosition.ParameterName = "@position";
                paramPosition.Value = item.Position;
                base.sqlCommand.Parameters.Add(paramPosition);

                var paramEmail = base.sqlCommand.CreateParameter();
                paramEmail.ParameterName = "@email";
                paramEmail.Value = item.Email;
                base.sqlCommand.Parameters.Add(paramEmail);

                var paramPassword = base.sqlCommand.CreateParameter();
                paramPassword.ParameterName = "@password";
                paramPassword.Value = item.Password;
                base.sqlCommand.Parameters.Add(paramPassword);

                return base.add(item);
            }
        }

        public override void delete(Employee item)
        {
            base.connection = DbUtils.getConnection(base.properties);
            using (base.sqlCommand = connection.CreateCommand())
            {
                base.sqlCommand.CommandText = "DELETE FROM employees WHERE id=@id";

                var paramId = base.sqlCommand.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = item.ID;
                base.sqlCommand.Parameters.Add(paramId);

                base.delete(item);
            }
        }

        public override void update(Employee item, int id)
        {
            base.connection = DbUtils.getConnection(base.properties);
            using (base.sqlCommand = connection.CreateCommand())
            {
                base.sqlCommand.CommandText = "UPDATE employees SET first_name=@firstName, last_name=@lastName, position=@position, " +
                    "email=@email, password=@password WHERE id=@id";

                var paramId = base.sqlCommand.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = item.ID;
                base.sqlCommand.Parameters.Add(paramId);

                var paramFirstName = base.sqlCommand.CreateParameter();
                paramFirstName.ParameterName = "@firstName";
                paramFirstName.Value = item.FirstName;
                base.sqlCommand.Parameters.Add(paramFirstName);

                var paramLastName = base.sqlCommand.CreateParameter();
                paramLastName.ParameterName = "@lastName";
                paramLastName.Value = item.LastName;
                base.sqlCommand.Parameters.Add(paramLastName);

                var paramPosition = base.sqlCommand.CreateParameter();
                paramPosition.ParameterName = "@position";
                paramPosition.Value = item.Position;
                base.sqlCommand.Parameters.Add(paramPosition);

                var paramEmail = base.sqlCommand.CreateParameter();
                paramEmail.ParameterName = "@email";
                paramEmail.Value = item.Email;
                base.sqlCommand.Parameters.Add(paramEmail);

                var paramPassword = base.sqlCommand.CreateParameter();
                paramPassword.ParameterName = "@password";
                paramPassword.Value = item.Password;
                base.sqlCommand.Parameters.Add(paramPassword);

                base.update(item, id);
            }
        }

        public override Employee? findById(int id)
        {
            base.connection = DbUtils.getConnection(base.properties);
            using (base.sqlCommand = connection.CreateCommand())
            {
                base.sqlCommand.CommandText = "SELECT * FROM employees WHERE id=@id";

                var paramId = base.sqlCommand.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                base.sqlCommand.Parameters.Add(paramId);

                return base.findById(id);
            }
        }

        public Employee? findByEmail(string email)
        {
            base.connection = DbUtils.getConnection(base.properties);
            using (base.sqlCommand = connection.CreateCommand())
            {
                base.sqlCommand.CommandText = "SELECT * FROM employees WHERE email=@email";

                var paramEmail = base.sqlCommand.CreateParameter();
                paramEmail.ParameterName = "@email";
                paramEmail.Value = email;
                base.sqlCommand.Parameters.Add(paramEmail);

                return base.getOne();
            }
        }

        public List<Employee> getByPosition(string position)
        {
            base.connection = DbUtils.getConnection(base.properties);
            using (base.sqlCommand = connection.CreateCommand())
            {
                base.sqlCommand.CommandText = "SELECT * FROM employees WHERE position=@position";

                var paramPosition = base.sqlCommand.CreateParameter();
                paramPosition.ParameterName = "@position";
                paramPosition.Value = position;
                base.sqlCommand.Parameters.Add(paramPosition);

                return base.findAll();
            }
        }

        public override List<Employee> findAll()
        {
            base.connection = DbUtils.getConnection(base.properties);
            using (base.sqlCommand = connection.CreateCommand())
            {
                base.sqlCommand.CommandText = "SELECT * FROM employees";

                return base.findAll();
            }
        }
    }
}
